### PR TITLE
Advertising topics assisted by ticketing

### PR DESCRIPTION
**Pull request by emhane**

### CODE REVIEW

**Review by divagant-martian — CHANGES_REQUESTED** [review]
So far I've only checked the advertisement, and left some comments. I'd like to hear @AgeManning opinion on using futures for getting the ticket expiration. Since part of the logic starts from the premise that the expire times are ordered, it could be enough to store the deadlines, and remove expired items on different operations. Using futures also makes the code subject to some subtle race conditions, in which re reject tickets because we see the table/queue full because poll has not been called to remove expired entries. I don't think this is a big deal at all, and would be fine with using the sleeps anyway

File: src/advertisement/mod.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    pub fn new(ad_lifetime: Duration, max_ads_per_topic: usize, max_ads: i32) -> Self {
+        Ads {
+            expirations: VecDeque::new(),
+            ads: HashMap::new(),
+            total_ads: 0,
+            ad_lifetime,
+            max_ads_per_topic,
+            max_ads,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs:58]: I set max_ads_per_topic and max_ads to default values now if the check fails. What action do you think we should take if the check fails? These parameters will eventually be set in the Discv5Config, so maybe the check isn't necessary here after all? @divagant-martian

[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/advertisement/mod.rs:58]: locally (as in wrt the Ads structure) I would return here a `Result<Self, &'static str>`
Then, when discovery starts we can `expect` on it. Discovery's strategy seems to be to panic on wrong configuration so that's what we would do following the rest of the code. Since the discovery config can be created without the need of the builder, the checks on the builder are not enough to guarantee things are right here

File: src/advertisement/mod.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                                return Err("No nodes in table".into());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn insert(&mut self, node_record: Enr<CombinedKey>, topic: Topic) -> Result<(), String> {
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs:118]: Aha, I should use Option because of this instead of Result and return None.

[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]:
```rust
topic_hash(topic).map(|topic| {
                    self.ads.ticket_wait_time(topic).map(|wait_time| {
                        let new_ticket = Ticket::new(topic);
                        self.send_ticket_response(
                            node_address.clone(),
                            id.clone(),
                            new_ticket,
                            wait_time,
                        );

                        let _ticket = Ticket::decode(ticket).unwrap_or(Ticket::default());

                        // use wait time to see if there is any point at doing ticket validation

                        // choose which ad to reg based on ticket, for example if some node has empty ticket
                        // or is coming back, and possibly other stuff

                        match self.ads.insert(enr, topic) {
                            Ok(()) => self.send_regconfirmation_response(node_address, id, topic),
                            Err(e) => error!("{}", e),
                        }
                    }).ok();
                }).ok();
```

This is what I had, it worked when the Error returned by ```topic_hash``` and ```ticket_wait_time``` was String, but not &str. At second thought, it would have worked with &'static str right?

[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]: Now I return Option, I think it makes more sense since I don't want to do anything if None is returned.

File: src/handler/mod.rs [hunk 3]
```diff
@@ -85,7 +85,7 @@ pub enum HandlerIn {
     ///
     /// Note: To update an ENR for an unknown node, we request a FINDNODE with distance 0 to the
     /// `NodeContact` we know of.
-    Request(NodeContact, Box<Request>),
+    Request(RequestContact, Box<Request>),
```
Comment (AgeManning) [inline src/handler/mod.rs:88]: I could be missing something here, but I dont think this is needed.

For the topic advertisement stuff, I was thinking we didn't need to touch the handler at all. The handler should handle sessions,encryption and all that kind of magic for us. I thought we could just modify the service and send different RPC's through. The idea being that the service doesn't need to worry about establishing sessions etc thats all handled in the `handler`. The `Handler` sends and manages `Requests`. It doesn't need to know really whats inside the requests, so I dont think needs a `regtopic_with_ticket()` function. 

I did miss a bit of the discussion of the problem we were facing with obtaining ENRs. This is how I thought it would work: 

Service decides a bunch of nodes to send RegTopic to. When it finds nodes, it must get their ENR (from local routing table) in order to send a request. 
Once it has selected a few ENRs, it sends a `Request` to the handler with RegTopic. As it stands, whether there is a session or not, the handler will establish a session if it needs to, or use a current established one. From a service perspective, we dont need to worry, handler deals with all that magic. Once the response comes back from the service's `RegTopic` request, it will be either a `Ticket` response or a `RegConfirmation` response (we can specify what allowed types we have here: https://github.com/sigp/discv5/blob/master/src/rpc.rs#L205). The handler could also spit out an error, indicating we couldn't establish a session or contact the node. 

Given a Response from the handler, we can get the ENR of the node that responded because we have `active_requests` in the `Service` (https://github.com/sigp/discv5/blob/master/src/service.rs#L163). 
So when the `Service` handles a response from the handler, it can get the `NodeContact`. See it here: https://github.com/sigp/discv5/blob/master/src/service.rs#L586

And really we just have to add a match arm here: https://github.com/sigp/discv5/blob/master/src/service.rs#L612 to handle the `Ticket` or `RegConfirmation` responses. And inside this match we have access to the ENR of the node we are getting a response from. 

The overall idea is that the handler hides all the complexity of sessions/encryption and all that. All we should have to modify is `service.rs` to handle the high level RPC's. `rpc.rs` to add the different rpc's we can send and how to encrypt/decrypt them. 

Let me know if i'm missing something tho

**Review by AgeManning — COMMENTED** [review]
Nice work. 

I didn't get all the way through this, but hopefully the parts I did get through the comments are helpful.

File: src/handler/mod.rs (hunk 3 above)
Comment (emhane) [reply src/handler/mod.rs:88]: This part is already implemented: "Service decides a bunch of nodes to send RegTopic to. When it finds nodes, it must get their ENR (from local routing table) in order to send a request."

The issue has been getting the ENR later for re-sending REGTOPIC when we got a TICKET response from the first REGTOPIC, and the ticket wait time has expired. I see your solution: upon receiving the TICKET response, extract the ENR from the active request and save it along with the ticket. I didn't pay attention to the ```active_requests``` in Service.

Thanks Age!

File: src/advertisement/ticket.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
+use super::*;
+use crate::node_info::NodeAddress;
+use delay_map::HashMapDelay;
+use std::cmp::Eq;
+use tracing::error;
+
+pub fn topic_hash(topic: Vec<u8>) -> Option<Topic> {
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/ticket.rs:7]: Exactly this is a placeholder.

File: src/advertisement/mod.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                    {
+                        error!("Mismatched mapping between expiration queue and total ads count");
+                        return None;
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn insert(&mut self, node_record: Enr<CombinedKey>, topic: Topic) -> Result<(), &str> {
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs:120]: Oh, now I see what you guys mean. Then this has to be applied to the function ```ticket_wait_time``` too, so both these operations will delete expired ads. My thought is to make ```insert``` private and only allow access through some ```regconfirmation``` function which compares the tickets of nodes applying for the ad place. This doesn't conflict with removing the Stream implementation.

File: src/advertisement/mod.rs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        self.node_record == other.node_record
+    }
+}
+pub struct Ads {
+    expirations: VecDeque<(Pin<Box<Sleep>>, Topic)>,
+    ads: HashMap<Topic, VecDeque<Ad>>,
+    total_ads: i32,
+    ad_lifetime: Duration,
+    max_ads_per_topic: usize,
+    max_ads: i32,
```
[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/advertisement/mod.rs]: Hey @emhane I think you are marking as resolved some comments that are unaddressed, as these here noted too by @AgeManning

[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]: This is addressed. Pretty sure it was pushed yesterday, I guess not.

File: src/advertisement/mod.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+};
+use tokio::time::{sleep, Instant, Sleep};
+use tracing::{debug, error};
+
+mod test;
+pub mod ticket;
+
+pub type Topic = [u8; 32];
+
+/// An ad we are adevrtising for another node
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]: Like this, I already committed this, I am certain of it @divagant-martian.

File: src/advertisement/mod.rs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+impl PartialEq for Ad {
+    fn eq(&self, other: &Self) -> bool {
+        self.node_record == other.node_record
+    }
+}
+pub struct Ads {
+    expirations: VecDeque<(Pin<Box<Sleep>>, Topic)>,
+    ads: HashMap<Topic, VecDeque<Ad>>,
+    total_ads: i32,
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]: @divagant-martian maybe it is because the comment is made on Outdated code?

File: src/advertisement/mod.rs [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+impl Ads {
+    pub fn new(
+        ad_lifetime: Duration,
+        max_ads_per_topic: usize,
+        max_ads: usize,
+    ) -> Result<Self, &'static str> {
+        let (max_ads_per_topic, max_ads) = if max_ads_per_topic <= max_ads {
+            (max_ads_per_topic, max_ads)
```
[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/advertisement/mod.rs:60]: Also thought about the builder but seemed too much for two params. That's why I suggested the error.
See https://github.com/sigp/discv5/pull/106#discussion_r844193934

File: src/handler/mod.rs (hunk 3 above)
Comment (emhane) [reply src/handler/mod.rs:88]: This I noticed however "it will be either a Ticket response or a RegConfirmation response" is contradicting the spec which, as I understood it states we should always issue a ticket. https://github.com/ethereum/devp2p/blob/master/discv5/discv5-wire.md#regtopic-request-0x07

File: src/advertisement/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]: I set this to a ```usize``` already, is that alright too?

[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/advertisement/mod.rs]: I think an `usize` it's correct :+1:

File: src/advertisement/mod.rs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    pub fn get_ad_nodes(&self, topic: Topic) -> Result<Iter<'_, Ad>, &str> {
+        match self.ads.get(&topic) {
+            Some(topic_ads) => Ok(topic_ads.into_iter()),
+            None => Err("No ads for this topic"),
+        }
+    }
+
+    pub fn ticket_wait_time(&self, topic: Topic) -> Option<Duration> {
+        let now = Instant::now();
+        if self.total_ads < self.max_ads {
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/mod.rs]: Correct, it is allowed. Thanks.

File: src/handler/mod.rs (hunk 3 above)
Comment (AgeManning) [reply src/handler/mod.rs]: Yes. You're right. This is interesting then. The RegConfirmation still references a requestid. 
So I think similar to the `FindNode` request, the RegTopic can get multiple responses for the same request. 

The logic there we need to be careful about. I.e we only hold requests for so long before they expire and we dont accept any more responses. Also we decide how long to wait for a RegConfirmation. I suspect we just wait a single RPC request timeout amount.

**Review by divagant-martian — COMMENTED** [review]
I haven't really finished reviewing since it's extensive. 

Some general suggestion about the code, on top of the specific ones:

1. Giving docs to most (if not all) new structures and fields, and if possible some comments in the code. Common things that most of the time require expressive names and docs:
    - Structures that contain booleans, numerical values and strings in place of a generic. For example:
`map: HashMap<String, u8>` could be improved to `name_count: HashMap<String, u8>`, further improved to 

        ```rust
        /// Count of seen people with the same name during the last 10 minutes
        name_count: HashMap<String, u8>
        ```

    - Intermediate structures that are, in a way, implementation details. A `Ticket`, `FindNodes` struct etc requires relatively little documentation because it's almost safe to assume whoever is reading the code is familiar with the spec, or at least they know where to find it. Everything in between needs docs, in this case for example structures like the `TicketHistory` and similar ones. Docs should try to explain how it's used/why it's needed/ what guarantees it provides.
2. Use constants for values that are used for some logic in the code. This helps with testing since if we change the constant there is not need to change the test. So for example the first snippet above should not be `in the last 10 minutes` but `in the last NAME_OF_CONSTANT`

In what the ticket systems respects, I think this is going in the right track. The topic radius is not used and needs to be estimated. Later it would be interesting to do some test on how this behaves

File: src/discv5.rs [hunk 11]
```diff
@@ -479,6 +488,93 @@ impl Discv5 {
         }
     }
 
+    // Use find_topic to find the Enrs the shortest XOR distance from the topic hash,
+    // and send the topic query to these nodes
+    pub fn topic_query_req(
+        &self,
+        enr: Enr,
```
Comment (AgeManning) [inline src/discv5.rs:495]: This ENR is supposed to be our ENR right?
We send our ENR to other nodes so they know to register it, correct?
If this is the case, I dont think we need to expose this as a parameter here. The service knows about our local enr, so could use just use it there without us having to pass it in. 
I saw that in responses, we reject ENR's that don't match the source, so putting someone else's here would just get rejected right?

File: src/discv5.rs [hunk 12]
```diff
@@ -479,6 +488,93 @@ impl Discv5 {
...
+                .await
+                .map_err(|e| RequestError::ChannelFailed(e.to_string()))?
+        }
+    }
+
+    // Use find_topic to find the Enrs the shortest XOR distance from the topic hash,
+    // and send the topic query to these nodes
+    pub fn reg_topic_req(
+        &self,
+        enr: Enr,
```
Comment (AgeManning) [inline src/discv5.rs:527]: Same here?

File: src/advertisement/topic.rs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.topic.cmp(&other.topic))
+    }
+}
+
+impl<H: Hasher> Ord for Topic<H> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.topic.cmp(&other.topic)
+    }
+}
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/topic.rs:174]: For sure, I had planned to add another field, with the string of the topic (before hashing) to make it easy for the human eye to compare topics would they have been hashed with different algorithms.

File: src/advertisement/ticket.rs [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    pub fn insert(&mut self, active_topic: ActiveTopic) -> Result<(), &str> {
+        self.remove_expired();
+        let count = self.ticket_cache.entry(active_topic).or_default();
+        if *count >= 3 {
+            error!("Max 3 tickets per (NodeId, Topic) accepted in 15 minutes");
+            return Err("Ticket limit reached");
+        }
+        *count += 1;
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/ticket.rs]: Good catch, thanks!

File: src/advertisement/mod.rs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        self.expirations
+            .iter()
+            .take_while(|ad| ad.insert_time.elapsed() >= self.ad_lifetime)
+            .for_each(|ad| {
+                let count = map.entry(ad.topic).or_default();
+                *count += 1;
+            });
+
+        map.into_iter().for_each(|(topic, index)| {
+            let entry_ref = self.ads.entry(topic).or_default();
```
[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/advertisement/mod.rs]: what I mean is that `or_default` should never be called. On one side, if we are removing things, we shouldn't be inserting keys with default values in general. And in this specific case, the key should exist given the way it was created. In practice, this code path will never be reached _as is_ right now. But leaving this here is prone to errors in future refactors.

TLDR: don't call `or_default` since here the purpose is not to ensure something is inserted

File: src/config.rs [hunk 16]
```diff
@@ -91,6 +91,8 @@ pub struct Discv5Config {
     /// will last indefinitely. Default is 1 hour.
     pub ban_duration: Option<Duration>,
 
+    pub topic_radius: Option<u64>,
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/config.rs]: Topic radius is WIP, for now I moved it to be set in the Discv5Config so it can be modified for testing.

File: src/advertisement/ticket.rs [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+                if reg_window.open_time.elapsed() > Duration::from_secs(10) {
+                    self.ticket_pools
+                        .remove_entry(&reg_window.topic)
+                        .map(|(topic, ticket_pool)| {
+                            self.expirations.pop_front();
+                            Poll::Ready(Some(Ok((topic, ticket_pool))))
+                        })
+                        .unwrap_or_else(|| Poll::Ready(Some(Err("Ticket selection failed".into()))))
+                } else {
+                    Poll::Pending
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/ticket.rs]: I changed it return PollPending if the code failed, and made the error message part of a debug_unreachable!(), did I understand you right?

File: src/advertisement/ticket.rs [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+
+    pub fn is_active_req(
+        &mut self,
+        req_id: RequestId,
+        node_id: NodeId,
+        topic: TopicHash,
+    ) -> Option<bool> {
+        self.remove_expired();
+        self.requests
+            .remove(&ActiveTopic::new(node_id, topic))
```
[the remark this answers is not in the document]
Comment (emhane) [reply src/advertisement/ticket.rs:251]: Good catch, should be a call to get not remove! Thanks!

File: src/discv5.rs (hunk 11 above)
Comment (emhane) [reply src/discv5.rs:495]: No the idea for this parameter is that the request is sent to a specific node. The local enr is added to the request on the Service layer. Here is some WIP code I used for testing, of a fork of yours :) https://github.com/emhane/discv5-cli/blob/2318f77f4017b2f9f35b5f91be11628c8e66c462/src/topic/mod.rs#L127

File: src/discv5.rs (hunk 12 above)
Comment (emhane) [reply src/discv5.rs:527]: No, here too the idea is to find out which nodes to send the requests to in a separate call to the discv5 instance. I did it like this for simplicity of doing the first integration testing. https://github.com/emhane/discv5-cli/blob/2318f77f4017b2f9f35b5f91be11628c8e66c462/src/topic/mod.rs#L210

File: src/advertisement/ticket.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/advertisement/ticket.rs]: you were using sleeps before, so we would need to schedule the next sleep. Now you are using a hashmap delay so there is no need to do anything else

**Comment by emhane** [comment]
Closing this PR as the protocol it implements has been replaced as of Devconnect Amsterdam 2022.
